### ISSUE TITLE
First is better than Last for Language Data Influence

### ISSUE THREAD

**Issue by bigshanedogg** [bigshanedogg/survey#28]
## Problem statement

1. classcification 모델 최적화에 도움이 되는 혹은 저해하는 데이터 examples를 판별하는 data influence measure TracIn-WE를 제안한다.
2. 기존 data influence measure의 한계인 computational complexity와 부족한 설명력을 개선한다.
  - first layer의 gradient를 사용해도 low-level, high-level information이 모두 반영될 수 있으며, 심지어 연산 비용이 더 저렴하다.

## Glossary
- data influence: $I(x, x')$, importance of training data $x$ to test data $x'$
  - data influence가 높다 -> $x$의 loss가 크며, 모델은 $x$와 $x'$을 유사하게 해석한다.
- cancellation effect: 샘플 사이의 influence magnitude sum이 influence sum보다 클 때 발생하며, 대부분의 샘플들의 influence magnitude를 높이고, data influence의 판별 능력을 감소시키는 효과
- case deletion diagnostics: train 샘플을 없애가며 test 샘플의 영향도를 파악하는 기법 (~1982년~)
  - proponents(positive influence인 샘플)을 지우면 test 샘플의 target class의 prob이 낮아지고 ( $DEL_+$ ), opponents(negative influence인 샘플)을 지우면 높아질 것이다 ( $DEL_-$ )

## Baseline
- 기존의 data influence measure는 모두 classcification 모델의 last-layer 사용을 가정
  - ![Screen Shot 2023-02-13 at 2 46 47 PM](https://user-images.githubusercontent.com/18084680/218380412-81ecf25f-8efd-44be-9666-befc41a40f79.png)
  - ![Screen Shot 2023-02-13 at 2 47 00 PM](https://user-images.githubusercontent.com/18084680/218380435-729213bb-1162-4bee-bbd6-d957e370f46e.png)
    - $\Delta{W_c}$가 작다 -> 해당 forward로 인한 ${W_c}$의 변화량이 작으며, 실제 모델 prediction에는 큰 영향을 주지 않았을 수 있다.
    - ${G(W)_c}$가 크다 -> ${W_c}$에 대한 gradient norm의 합이 크다.
    - $C(W)$는 cancellation ratio를 의미하며, ${W_c}$에 대한 gradient norm은 크지만 실제 업데이트는 적을 때 높아진다.
- $I(x, x')$ = $L(x)S(x,x')L(x')$의 꼴을 가지고 있음
- vocab_size v, word_embedding의 dim이 d이며, m개의 training examples의 n개의 test examples에 대한 data influence 측정 시 복잡도: $O(mn{d^2}{v^2}+{d^3}{v^3})$
  - Influence Functions: ![Screen Shot 2023-02-13 at 3 03 16 PM](https://user-images.githubusercontent.com/18084680/218382640-c9be0553-8d2b-49ba-9f9c-d8593e5e253a.png)
  - Representer Points: $L(x)S(x,x')$
  - TracIN: $L(x)S(x,x')L(x')$

## Data details
- Toxicity
- AGnews
- MNLI

## Approach
- ![Screen Shot 2023-02-13 at 3 01 03 PM](https://user-images.githubusercontent.com/18084680/218382354-94f465ec-3eed-41db-b530-27a306ac2a25.png)
- last-layer는 데이터 포인트의 단어들 간의 정보가 섞인 후이기 때문에 'share-logic'으로 볼 수 있고, first-layer는 임베딩 최초의 값이기 때문에 i step의 벡터가 각 단어 별 영향도 정보가 희석되지 않은 'unique-logic'으로 볼 수 있다.
- last-layer를 사용할 경우, share-logic이기 때문에 i번째 벡터가 특정 단어의 영향도를 대변한다고 볼 수 없기 때문에, 전체 매트릭스를 사용해야 하며, 이는 계산량 증가를 가져온다. 또한, 불필요한 특정 단어가 포함되면서 data influence 특정에도 좋지 않다.
- first-layer는 unique-logic이기 때문에 i step이 i번째 단어의 영향도를 대변한다고 볼 수 있으며, 따라서 $L(x)S(x,x')L(x')$ 중 $S(x,x')$를 전체 벡터가 아닌, train example과 test example의 overlapping words에 대해서만 계산할 수 있다. 

- first-layer를 사용하면 low-level information만 반영되지 않을까? 특정 observation과 syntactic 구조에 영향을 많이 받을 것 같다
  - first-layer에는 high-level input semantics가 없어서 data influence 효과를 감지하는 것이 어렵지 않을까 하는 우려가 있었으나, last-layer로부터 오는 gradient chain의 영향 때문에 first-layer도 유의미하게 high-level information, 즉 다른 단어들과의 contextualized information이 내포된 것으로 보인다.
  - 다만, 첫째 layer는 단어 구성에 영향을 많이 받기 때문에, overlapping words가 많지만 semantic이 유사하지 않은 샘플들이나, 그 반대의 경우에 취약한 면이 있다.
- train data와 test data의 word overlapping이 없는 경우에도 유의미하게 동작하는지?
  - 최근 NLP model의 경우, 공통된 special token이 들어간다. 앞서 언급한 high-level information이 반영되기 때문에 이 두 토큰으로 인해 유의미한 결과가 나온다. TFIDF
  - 

## Evaluation
- deletion curve로 평가 (case deletion diagnostics 관련)
  - proponents 그룹과 opponents 그룹을 k개씩 제거하면서 baseline과의 ${DEL_+}$와 ${DEL_-}$을 측정
  - ![Screen Shot 2023-02-13 at 2 41 53 PM](https://user-images.githubusercontent.com/18084680/218379774-0535583e-2c9c-443a-8137-f88e986e8948.png)
  - ${DEL_+}(x', k, I)$ (proponents를 지웠을 때 target class prob의 mean)이 작을수록, ${DEL_-}(x', k, I)$ (opponents를 지웠을 때 target class prob의 mean)이 클수록, 좋은 Influence measure이다.
  - ![Screen Shot 2023-02-13 at 2 40 56 PM](https://user-images.githubusercontent.com/18084680/218379640-f31c27c5-0259-4a42-8f71-92f5c6dc74f2.png)
  - top-k word만 사용하는 TracIn-WE-top-k의 경우 더 많은 연산비용 절감이 가능한데 비해, 성능도 크게 떨어지지 않음을 확인할 수 있다.

## Limitations
- 생각해봅시다.
